File: Android/service/src/main/java/com/scurab/android/anuitor/nanoplugin/ActivityPlugin.java
package com.scurab.android.anuitor.nanoplugin;

import android.app.Activity;
import android.view.View;

import com.scurab.android.anuitor.reflect.WindowManager;
import com.scurab.android.anuitor.tools.HttpTools;

import java.io.File;
import java.util.HashMap;
import java.util.Map;

import fi.iki.elonen.NanoHTTPD;

/**
 * User: jbruchanov
 * Date: 12/05/2014
 * Time: 15:53
 */
public abstract class ActivityPlugin extends BasePlugin {

    protected static final String POSITION = "position";
    public static final String APPLICATION_IS_NOT_ACTIVE = "Application is not active";
    private static final String SCREEN_INDEX = "screenIndex";

    private static final NanoHTTPD.Response EMPTY_RESPONSE = new OKResponse(HttpTools.MimeType.TEXT_PLAIN, APPLICATION_IS_NOT_ACTIVE);
    private static final NanoHTTPD.Response JSON_EMPTY_RESPONSE = new OKResponse(HttpTools.MimeType.APP_JSON, "{}");

    private WindowManager mWindowManager;

    protected ActivityPlugin(WindowManager windowManager) {
        if (windowManager == null) {
            throw new IllegalArgumentException("knowActivity is null!");
        }
        mWindowManager = windowManager;
    }

    @Override
    public final NanoHTTPD.Response serveFile(String uri, Map<String, String> headers, NanoHTTPD.IHTTPSession session, File file, String mimeType) {
        String[] viewRootNames = mWindowManager.getViewRootNames();
        if (viewRootNames == null || viewRootNames.length == 0) {
            if (mimeType.equals(HttpTools.MimeType.APP_JSON)) {
                return JSON_EMPTY_RESPONSE;
            } else {
                return EMPTY_RESPONSE;
            }
        } else {
            return handleRequest(uri, headers, session, file, mimeType);
        }
    }

    public abstract NanoHTTPD.Response handleRequest(String uri, Map<String, String> headers, NanoHTTPD.IHTTPSession session, File file, String mimeType);

    public Activity getCurrentActivity() {
        return mWindowManager.getCurrentActivity();
    }

    public View getCurrentRootView() {
        return getCurrentRootView(-1);
    }

    public View getCurrentRootView(int index) {
        return index < 0 ? mWindowManager.getCurrentRootView() : mWindowManager.getRootView(index);
    }

    public View getCurrentRootView(HashMap<String, String> qsValue) {
        View view;
        try {
            if (qsValue.containsKey(SCREEN_INDEX)) {
                String index = qsValue.get(SCREEN_INDEX);
                int i = Integer.parseInt(index);
                view = getCurrentRootView(i);
            } else {
                view = getCurrentRootView();
            }
        } catch (Throwable t) {
            t.printStackTrace();
            view = null;
        }
        return view;
    }
}

File: Android/service/src/main/java/com/scurab/android/anuitor/nanoplugin/ActivityPlugin.kt
package com.scurab.android.anuitor.nanoplugin

import android.app.Activity
import android.view.View
import com.scurab.android.anuitor.reflect.WindowManager
import com.scurab.android.anuitor.tools.HttpTools
import fi.iki.elonen.NanoHTTPD
import java.io.File

abstract class ActivityPlugin protected constructor(private val windowManager: WindowManager) : BasePlugin() {

    val currentActivity: Activity
        get() = windowManager.currentActivity

    val currentRootView: View
        get() = getCurrentRootView(-1)

    override fun serveFile(uri: String, headers: Map<String, String>, session: NanoHTTPD.IHTTPSession, file: File, mimeType: String): NanoHTTPD.Response {
        val viewRootNames = windowManager.viewRootNames
        return if (viewRootNames == null || viewRootNames.isEmpty()) {
            if (mimeType == HttpTools.MimeType.APP_JSON) {
                OKResponse(HttpTools.MimeType.APP_JSON, "[]")
            } else {
                OKResponse(HttpTools.MimeType.TEXT_PLAIN, APPLICATION_IS_NOT_ACTIVE)
            }
        } else {
            onRequest(uri, headers, session, file, mimeType)
        }
    }

    abstract fun onRequest(uri: String, headers: Map<String, String>, session: NanoHTTPD.IHTTPSession, file: File, mimeType: String): NanoHTTPD.Response

    fun getCurrentRootView(index: Int): View {
        return if (index < 0) windowManager.currentRootView else windowManager.getRootView(index)
    }

    protected fun String.currentRootView(): View? {
        return try {
            this
                    .let { HttpTools.parseQueryString(it) }
                    .let { it.currentRootView() }
        } catch (e: Throwable) {
            null
        }
    }

    protected fun Map<String, String>.currentRootView(): View? {
        return try {
            return this
                    .let { (it[SCREEN_INDEX] ?: "-1").toInt() }
                    .let { getCurrentRootView(it) }
        } catch (e: Throwable) {
            null
        }
    }

    companion object {
        const val POSITION = "position"
        const val SCREEN_INDEX = "screenIndex"
        const val APPLICATION_IS_NOT_ACTIVE = "Application is not active"
    }
}
